action/action-util/src/main/java/com/powsybl/action/util/ProportionalScalable.java

```diff
@@ -25,30 +99,65 @@ private static void checkPercentages(List<Float> percentages, List<Scalable> sca
         }
         double sum = percentages.stream().mapToDouble(Double::valueOf).sum();
         if (Math.abs(100 - sum) > 0.01) {
-            throw new IllegalArgumentException("Sum of percentages must be equals to 100 (" + sum + ")");
+            throw new IllegalArgumentException(String.format("Sum of percentages must be equals to 100 (%.2f)", sum));
         }
     }
 
-    private final List<Float> percentages;
+    private boolean notSaturated() {
+        return scalablePercentageList.stream().anyMatch(ScalablePercentage::notSaturated);
+    }
 
-    ProportionalScalable(List<Float> percentages, List<Scalable> scalables) {
-        super(scalables);
-        checkPercentages(percentages, scalables);
-        this.percentages = Objects.requireNonNull(percentages);
+    private void checkIterationPercentages() {
+        double iterationPercentagesSum = scalablePercentageList.stream().mapToDouble(ScalablePercentage::getIterationPercentage).sum();
+        if (Math.abs(100 - iterationPercentagesSum) > 0.01) {
+            throw new AssertionError(String.format("Error in proportional scalable ventilation. Sum of percentages must be equals to 100 (%.2f)", iterationPercentagesSum));
+        }
+    }
+
+    private void updateIterationPercentages() {
+        double unsaturatedPercentagesSum = scalablePercentageList.stream().filter(ScalablePercentage::notSaturated).mapToDouble(ScalablePercentage::getIterationPercentage).sum();
+        scalablePercentageList.forEach(scalablePercentage -> {
+            if (scalablePercentage.isSaturated()) {
+                scalablePercentage.setIterationPercentage(0);
+            } else {
+                scalablePercentage.setIterationPercentage(scalablePercentage.getPercentage() / unsaturatedPercentagesSum * 100);
+            }
+        });
+    }
+
+    private double iterativeScale(Network n, double asked, ScalingConvention scalingConvention) {
+        double done = 0;
+        while (Math.abs(asked - done) > 1e-5 && notSaturated()) {
```

Comment:
You should create a constant for epsilon

Reply:
done